### PR TITLE
privacy: Fix regression in impl reachability

### PR DESCRIPTION
**Pull request by petrochenkov**
Rollback to pre-https://github.com/rust-lang/rust/pull/56878 logic of determining reachability.
`reachability(impl Trait<Substs> for Type<Substs>) = reachability(Trait & Type)`, substs are ignored.

Fixes https://github.com/rust-lang/rust/issues/57264

### CODE REVIEW

**Comment by arielb1** [comment]
I think the invariant the impl rule tries to rely on should be documented somewhere:

The invariant is that type-inference won't let you use an impl without knowing the "shallow" version of its self type (which requires reaching the `DefId`s in it).

I noted this is not handled correctly with projections, so I have an example that ICEs even older rustcs:

aux.rs:
```rust
mod inner {
    #[derive(Default)]
    pub struct PubUnnameable;
}

trait Mirror { type Image; }
impl<T> Mirror for T { type Image = T; }

pub enum Pub<T> { Nothing, Just(T) }
pub trait Aux {}
impl Aux for <Pub<inner::PubUnnameable> as Mirror>::Image {}

pub fn assert_aux<T: Aux>(_t: &T) {}

impl inner::PubUnnameable {
    pub fn pub_method(self) {}
}
```

main.rs:
```rust
extern crate aux;

fn get_default<T: Default>(_t: &aux::Pub<T>) -> T { T::default() }

fn main() {
    let pub_ = aux::Pub::Nothing;
    aux::assert_aux(&pub_);
    let def = get_default(&pub_);
    def.pub_method();
}
```

**Comment by arielb1** [comment]
I would rename `RECURSE` into `!SHALLOW`, and make sure that the few other recursive calls in `visit_ty` don't occur in shallow mode.

**Comment by arielb1** [comment]
r=me with these changes

File: src/librustc_privacy/lib.rs [hunk 1]
```diff
@@ -206,7 +208,7 @@ impl<'a, 'tcx, V> TypeVisitor<'tcx> for DefIdVisitorSkeleton<'_, 'a, 'tcx, V>
                 bug!("unexpected type: {:?}", ty),
```
Comment (arielb1) [inline src/librustc_privacy/lib.rs]: There are a few other cases where `visit_ty` can recurse, you need to avoid them.

**Comment by petrochenkov** [comment]
>I noted this is not handled correctly with projections, so I have an example that ICEs even older rustcs

Projections are not necessary in that case, the root reason is a public trait that's implemented for a single type, so type inference can make conclusion "`T: Trait` => `T = ThatType`".
https://github.com/rust-lang/rust/pull/57344/commits/9503c56ff7cbbc04228fbca6fde58d81ec4d3d87 has a further minimized example and a fix.

@bors r=arielb1

**Comment by arielb1** [comment]
> > I noted this is not handled correctly with projections, so I have an example that ICEs even older rustcs
> 
> Projections are not necessary in that case, the root reason is a public trait that's implemented for a single type, so type inference can make conclusion "`T: Trait` => `T = ThatType`".
> [9503c56](https://github.com/rust-lang/rust/commit/9503c56ff7cbbc04228fbca6fde58d81ec4d3d87) has a further minimized example and a fix.

Ok. And there's no inherent problem with projections because accessibility computation has to treat projections as public anyway.

> @bors r=arielb1

I would prefer you to not r=me after you make a change this size.

In any case:

Looking at the code, the one point that I think is missing an adequate explanation is the `Ty::Opaque` case where you are recurring into the contents. However, I think this is currently OK because:
1. I don't think we can have `Ty::Opaque` in the self-type of an impl, and if that becomes possible (`abstract type`), we'll have to rethink the privacy of `pub abstract type MyType: PublicTrait + PrivateTrait` anyway.
2. In any case, the privacy property we are checking is the privacy of the *def id*, not the privacy of the substs. If a user can get his hands on an instance of an `Opaque`, "inductively on the type-checking of the program" we can know that the def id of that `Opaque` is not private, which means that we won't mark it as private in our pass.

However, I think that is remarkable enough I'll want it as a comment.

@bors r-

**Comment by arielb1** [comment]
r=me with a comment on `ty::Opaque` (it would be OK to copy what I wrote and put it in a comment).

(If you are only changing the comment, I'm ok with a `@​bors r=arielb1`).

**Comment by petrochenkov** [comment]
@arielb1 
From privacy point of view `impl Trait1 + Trait2` should be checked identically to `dyn Trait1 + Trait2`, so we are not even interested in the def-id of the opaque itself (which either doesn't have attached visibility or it's insignificant, like visibility of a type alias).
It's private if any trait in the list is private.

This intent is somewhat hidden because impl-Trait and dyn-Trait use such different structures in `ty` and we have to recurse to get the trait list in one case, but not in another.

**Pull request merged**